### PR TITLE
websocket bridge and browser web client implementations

### PR DESCRIPTION
**Pull request by JosePedroDias**
I tried to make your server work with websockets as well and I think it's usable. Haven't delved into websocket server config tweaking but seems promising.
Updated the README as well, with my formatter changing the file substantially. Let me know if you want me to remove/revise that file. Cheers!

P.S. - if updating the coding style is a problem I can rollback most of node.js file changes and just prepare the ws hook instead.

Now that I think of it, I made a small optimization where the socket doesn't get its own messages back. That breaks the lag estimation example. Let me know if that's a problem, may be turned into an optional config param.

### CODE REVIEW

**Comment by dr3am3r** [comment]
Also, please make sure tests make sense, currently I see this in upstream master:
Screenshot 2020-06-02 at 23 18 58
In comparison with yours:
Screenshot 2020-06-02 at 23 19 57

File: client/javascript-node-js/client.js [hunk 1]
```diff
@@ -3,105 +3,120 @@
...
+  const self = this;
 
-  this.socket = null
-  this.buffer = new Buffer(1024 * 16)
-  this.buffer.len = 0
-  this.messageCallback = null
-  this.errorCallback = null
-  this.subscribedCallback = null
+  this.socket = null;
+  this.buffer = new Buffer.alloc(1024 * 16);
```
[the remark this answers is not in the document]
Comment (JosePedroDias) [reply client/javascript-node-js/client.js:24]: you're right, kept it by mistake.

File: client/javascript-node-js/client.js [hunk 2]
```diff
@@ -3,105 +3,120 @@
...
+function _log() {
+  if (VERBOSE) console.log.apply(console, arguments);
+}
+
+const configDef = {
   server: 'localhost',
   port: 1337,
-  channel: 'gsom'
-}
+  channel: 'gsom',
```
[the remark this answers is not in the document]
Comment (JosePedroDias) [reply client/javascript-node-js/client.js:17]: fixing it

File: client/javascript-browser/client.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    const ws = new WebSocket(`ws://${cfg.server}:${cfg.port}`);
+
+    let channel;
+    let callback;
+    let errorCallback;
+    let subscribedCallback;
+    let isReady = false;
+
+    function publish(data) {
+      isReady && ws.send(`__JSON__START__${JSON.stringify(data)}__JSON__END__`);
```
[the remark this answers is not in the document]
Comment (JosePedroDias) [reply client/javascript-browser/client.js:14]: I agree. tried to use the same logic as the main tcp server whenever possible. due to the atomic messages ws receives, I could drop some buffering split/join logic. These can be extracted yeah.

File: README.md [hunk 4]
```diff
@@ -1,39 +1,47 @@
...
-* Connections are routed through socket server with minimum latency, ideal for action games.
-* Simple interface. Publish/subscribe paradigm in action.
-* Server written on blazing fast Nodejs.
-* Zero dependency. Works out of the box, no NPM ecosystem required.
-* Socket connections, works great through any NAT (local area network), messages delivery is reliable and fast.
-* Low CPU and memory footprint
+- Connections are routed through socket server with minimum latency, ideal for action games.
+- Simple interface. Publish/subscribe paradigm in action.
+- Server written on blazing fast Nodejs.
+- Zero dependency. Works out of the box, no NPM ecosystem required.
```
[the remark this answers is not in the document]
Comment (JosePedroDias) [reply README.md:12]: There is if you use the bridge to ws. I wasn't sure how to address this... is it a peer dependency? (I didn't want to assume this optional feature to be a requirement but it's your call)

**Comment by JosePedroDias** [comment]
> Also, please make sure tests make sense, currently I see this in upstream master:
> Screenshot 2020-06-02 at 23 18 58
> In comparison with yours:
> Screenshot 2020-06-02 at 23 19 57

Regarding the `run-tests.sh` bash, it doesn't work on mac the same way as where you're probably testing (linux? mingw?) can you advise?

The reason why the test fails is because it assumes the client gets its own messages back. I thought it would be best not to do that (less message passing and less processing at each client) but one loses ping capabilities. I can either drop that optimization or make it optional...

File: client/javascript-node-js/client.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (JosePedroDias) [reply client/javascript-node-js/client.js]: done

File: client/javascript-browser/client.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (JosePedroDias) [reply client/javascript-browser/client.js:14]: Haven't done this yet. If you want to keep client and server self sufficient, i would have to define the consts on both. If I extract both to the server, the client needs to require it. 🤔

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (JosePedroDias) [reply README.md]: I mentioned it.

**Comment by JosePedroDias** [comment]
The tests should work now because I made the previous behaviour default. Clients get their own messages back unless default config is toggled off.

**Pull request merged**